### PR TITLE
Setup ol-web1 yet again

### LINKED ISSUE #3783: Create ol-web1
Fossa + docker (cc @samuel-archive )

### PR DESCRIPTION
**Pull request by cclauss**
Closes #3880, #3901 
Closes #3783

Allow us to convert `ol-web1` with a fresh Ubuntu install into an Open Library web node running Python 3.8 on Docker.

The workflow will be:
1. Log into ol-web1 that has a fresh Ubuntu install.
2. Get nginx is running OUTSIDE of Docker
    1. Use github token to git clone github.com/internetarchive/olsystem into /olsystem
    2. Make symbolic links to /etc/nginx/nginx.conf and /etc/nginx/sites-available
    3. Copy ol-web3:/opt/.petabox/seed to ol-web1:/opt/.petabox/seed
    4. sudo systemctl start nginx
3. `systemctl status nginx ` # to ensure it is up-and-running 
4. Copy `scripts/setup_web1.sh` into your home directory.
5. Set environment variables and run `setup_web1.sh` to and configure docker, openlibrary, olsystem
    * `export GITHUB_USERNAME=myusername OL_DOMAIN=myusername OL_BRANCH=Setup-ol-web1-yet-again`
    * `export GITHUB_TOKEN=mytoken ` # This ___required___ to git clone olsystem
    * `./setup_web1.sh` which will run `docker-compose.production.yml`
    * `docker-compose.production.yml` will run `docker/ol-web-start.sh`.
6. __BUG__ the last lines of `scripts/setup_web1.sh` need to be run manually...
    * `systemctl status nginx && cd /opt/openlibrary`
    * `sudo docker-compose -f docker-compose.yml -f docker-compose.infogami-local.yml \`
                                       `-f docker-compose.production.yml up --no-deps -d web`
    * `sudo docker-compose logs --tail=100 -f web`
    * OPTIONAL: bash in with `sudo docker exec -it openlibrary_web_1 /bin/bash`

### Testing

___This dangerous code should only be run in non-production environments.  It restores `ol-web1` to a clean state.___
```bash
#!/bin/bash

cd /opt/openlibrary
sudo docker-compose down
sudo systemctl disable docker
sudo systemctl stop docker
sudo docker system prune -y
sudo apt-get --purge remove -y docker docker.io docker-compose
sudo rm -fr /opt/containerd /opt/openlibrary /opt/olsystem
ls -l /opt  # nothing
```

### CODE REVIEW

File: docker/ol-web-start.sh [hunk 1]
```diff
@@ -1,5 +1,13 @@
 #!/bin/bash
 
+if [[ "$USE_NGINX" ]]; then
+  mkdir /etc/nginx || true
+  rm -f /etc/nginx/nginx.conf || true
+  ln -sf /opt/olsystem/etc/nginx/nginx.conf      /etc/nginx/nginx.conf
+  ln -sf /opt/olsystem/etc/nginx/sites-available /etc/nginx/sites-available
```
Comment (cdrini) [inline docker/ol-web-start.sh:7]: Ah, this shouldn't be `/opt/olsystem`; that path doesn't exist inside the docker container.

```suggestion
  ln -sf /olsystem/etc/nginx/nginx.conf      /etc/nginx/nginx.conf
  ln -sf /olsystem/etc/nginx/sites-available /etc/nginx/sites-available
```

File: scripts/setup_web1.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+sudo systemctl enable docker
+
+sudo groupadd --system openlibrary
+sudo useradd --no-log-init --system --gid openlibrary --create-home openlibrary
+
+cd /opt
+ls -l  # nothing
+
+# sudo git clone https://github.com/internetarchive/openlibrary
+sudo git clone https://github.com/cclauss/openlibrary
```
Comment (cdrini) [inline scripts/setup_web1.sh:26]: Maybe this?

```suggestion
OL_REPO=${OL_REPO:-internetarchive}
sudo git clone https://github.com/$OL_REPO/openlibrary
```

File: scripts/setup_web1.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
```
Comment (cdrini) [inline scripts/setup_web1.sh:2]:
```suggestion
# This script is used to provision an ol-webX node _before_ docker gets on it.
# It git clones the repos, installs docker, creates openlibrary user, etc.

```

File: scripts/setup_web1.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+cd /opt
+ls -l  # nothing
+
+# sudo git clone https://github.com/internetarchive/openlibrary
+sudo git clone https://github.com/cclauss/openlibrary
+sudo git clone https://${GITHUB_USERNAME:-$USER}:${GITHUB_TOKEN}@github.com/internetarchive/olsystem
+# sudo git pull https://${GITHUB_USERNAME:-$USER}:${GITHUB_TOKEN}@github.com/internetarchive/olsystem.git master
+sudo chown openlibrary /opt/*
+ls -l  # containerd, olsystem, openlibrary owned by openlibrary
```
Comment (cdrini) [inline scripts/setup_web1.sh:30]: What's `containerd`?

File: docker/ol-web-start.sh (hunk 1 above)
Comment (cclauss) [reply docker/ol-web-start.sh]: Does that mean that I do not need to clone in olsystem?

File: scripts/setup_web1.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+cd /opt/openlibrary
+sudo git checkout Setup-ol-web1-yet-again
```
Comment (cdrini) [inline scripts/setup_web1.sh:33]:
```suggestion
OL_BRANCH=${OL_BRANCH:-master}
sudo git checkout "${OL_BRANCH}"
```

File: openlibrary/plugins/upstream/addbook.py [hunk 6]
```diff
@@ -639,13 +637,25 @@ def process_work(self, work):
...
-
-            f = StringIO(subjects.encode('utf-8'))  # no unicode in csv module
+            if six.PY2:
+                subjects = subjects.encode('utf-8')  # no unicode in csv module
+            f = six.StringIO(subjects)
             dedup = set()
             for s in next(csv.reader(f, dialect='excel', skipinitialspace=True)):
-                s = s.decode('utf-8')
+                if six.PY2:
+                    s = s.decode('utf-8')
```
Comment (cdrini) [inline openlibrary/plugins/upstream/addbook.py:658]: Were these meant to be part of this PR? if they're fixing Py2/3 bugs, let's pull them into a separate PR so we can test to makes sure they work on dev.openlibrary.org with Py2.

**Review by cdrini — CHANGES_REQUESTED** [review]
Looking good! A few small changes.

File: scripts/setup_web1.sh (hunk 4 above)
Comment (cclauss) [reply scripts/setup_web1.sh:30]: tl;dr An essential building block for Docker and K8
* https://www.docker.com/blog/what-is-containerd-runtime/
* https://containerd.io/

Comment (cdrini) [reply scripts/setup_web1.sh:30]: Ahh; how did it get here? Did you have to check it out, or is it something that gets added automatically somehow?

Comment (cclauss) [reply scripts/setup_web1.sh:30]: https://github.com/internetarchive/openlibrary/pull/3938/files#diff-3b30bef8970e22874431ef6aa3d67073bddc40ba78dc7e5a6c86e549a15b5706R13

File: openlibrary/plugins/upstream/addbook.py (hunk 6 above)
Comment (cdrini) [reply openlibrary/plugins/upstream/addbook.py]: DONE

File: scripts/setup_web1.sh (hunk 5 above)
Comment (cdrini) [reply scripts/setup_web1.sh]: DONE

File: scripts/setup_web1.sh (hunk 3 above)
Comment (cdrini) [reply scripts/setup_web1.sh]: DONE

File: scripts/setup_web1.sh (hunk 2 above)
Comment (cdrini) [reply scripts/setup_web1.sh]: DONE

**Comment by cdrini** [comment]
(I went through and "DONE"'d and resolved my comments; for future reference it would great if you could do that :) )

File: docker-compose.production.yml [hunk 7]
```diff
@@ -11,6 +11,10 @@ services:
       - GUNICORN_OPTS= --workers 50 --timeout 300  # 5 minutes before reload
       - OL_CONFIG=/olsystem/etc/openlibrary.yml
       - PYENV_VERSION=3.8.6
-      - USE_NGINX=true
     volumes:
       - ../olsystem:/olsystem
+    command: >
+      bash -c "ls -Fla /etc/nginx
+      && ln -sfv /olsystem/etc/nginx/nginx.conf /etc/nginx/nginx.conf
+      && ln -sfv /olsystem/etc/nginx/sites-available /etc/nginx/sites-available
```
Comment (cclauss) [inline docker-compose.production.yml:19]: I am just spinning my wheels on this.  I can not `rm` or `chown` or `sudo` the destination files so I can not create the symlinks.  What else should I try?

**Review by cdrini — CHANGES_REQUESTED** [review]
We tested and it looks like nginx is no longer needed on webnodes \o/ So let's remove those sections.

**Comment by cclauss** [comment]
Video https://archive.org/details/setup-of-ol-web-1 demonstrates this code in action.

File: scripts/setup_web1.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+# Set $GITHUB_USERNAME or $USER will be used.
+# Set $GITHUB_TOKEN or this script will halt.
+if [[ -z ${GITHUB_TOKEN} ]]; then
+    echo "FATAL: Can not git clone olsystem" ;
+    exit 1 ;
+fi
+
+# apt list --installed
+sudo apt-get update
+sudo apt-get install -y docker.io docker-compose nginx
```
Comment (cdrini) [inline scripts/setup_web1.sh:14]:
```suggestion
sudo apt-get install -y docker.io docker-compose
```

**Review by cdrini — APPROVED** [review]
Lgtm; removing the last reference to nginx. Note I haven't tested running this.
